sdk/src/main/java/com/amplitude/experiment/DefaultExperimentClient.kt

```diff
@@ -367,17 +372,265 @@ internal class DefaultExperimentClient internal constructor(
+            variant = fallbackVariant,
+            source = VariantSource.FALLBACK_CONFIG,
+            hasDefaultVariant = defaultVariantAndSource.hasDefaultVariant
+        )
+        if (!fallbackVariant.isNullOrEmpty()) {
+            return fallbackVariantAndSource
+        }
+        return defaultVariantAndSource
+    }
+
+    /**
+     * For Source.InitialVariants, fallback order goes:
+     *
+     *  1. Initial variants
+     *  2. Local storage
+     *  3. Inline function fallback
+     *  4. Config fallback
+     *
+     * If there is a default variant and no fallback, return the default variant.
+     */
+    private fun initialVariantsVariantAndSource(
+        key: String,
+        fallback: Variant? = null
+    ): VariantAndSource {
+        var defaultVariantAndSource = VariantAndSource()
+        // Initial variants
+        val initialVariantsVariant = config.initialVariants[key]
+        if (initialVariantsVariant != null) {
+            return VariantAndSource(
+                variant = initialVariantsVariant,
+                source = VariantSource.INITIAL_VARIANTS,
+                hasDefaultVariant = false
+            )
+        }
+        // Local storage
+        val localStorageVariant = variants.getAll()[key]
+        val isLocalStorageDefault = localStorageVariant?.metadata?.get("default") as? Boolean
+        if (localStorageVariant != null && isLocalStorageDefault != true) {
+            return VariantAndSource(
+                variant = localStorageVariant,
+                source = VariantSource.LOCAL_STORAGE,
+                hasDefaultVariant = false
+            )
+        } else if (isLocalStorageDefault == true) {
+            defaultVariantAndSource = VariantAndSource(
+                variant = localStorageVariant,
+                source = VariantSource.LOCAL_STORAGE,
+                hasDefaultVariant = true
+            )
+        }
+        // Inline fallback
+        if (fallback != null) {
+            return VariantAndSource(
+                variant = fallback,
+                source = VariantSource.FALLBACK_INLINE,
+                hasDefaultVariant = defaultVariantAndSource.hasDefaultVariant
+            )
+        }
+        // Configured fallback, or default variant
+        val fallbackVariant = config.fallbackVariant
+        val fallbackVariantAndSource = VariantAndSource(
+            variant = fallbackVariant,
+            source = VariantSource.FALLBACK_CONFIG,
+            hasDefaultVariant = defaultVariantAndSource.hasDefaultVariant
+        )
+        if (fallbackVariant.isNullOrEmpty()) {
+            return fallbackVariantAndSource
+        }
+        return defaultVariantAndSource
+    }
+
+
 }
 
-data class VariantAndSource(val variant: Variant, val source: VariantSource)
+data class VariantAndSource(
+    val variant: Variant = Variant(),
+    val source: VariantSource = VariantSource.FALLBACK_CONFIG,
+    val hasDefaultVariant: Boolean? = null
```

Comment:
You should be able to make this non-null (default false), which should simplify the rest of the code quite a bit.